### PR TITLE
Make cross-importing faster

### PR DESCRIPTION
**Pull request by beccadax**
We previously computed cross-imports by comparing N transitive imports against N transitive imports. This is wasteful, because at least one of the two modules in a pair has to actually declare a cross-import overlay for us to discover one, and the vast majority of modules don’t declare any.

This commit makes us instead compare N transitive imports against M transitive imports which are known to declare at least one cross-import overlay. Since N is potentailly in the thousands while M is perhaps in the double digits, this should be good for a substantial time savings.

However, this optimization has made a test of another cross-import performance optimization fail—not because we have regressed on that, but because it skips work the test case expects us to perform. I have XFAILed that test for now in case someone needs to merge this while I'm on vacation; if that doesn't happen, I'll fix that problem and merge this when I get back.

Fixes <rdar://problem/59538458>.

### CODE REVIEW

**Comment by brentdax** [comment]
@swift-ci please test

File: include/swift/AST/Module.h [hunk 1]
```diff
@@ -260,6 +260,11 @@ class ModuleDecl : public DeclContext, public TypeDecl {
   /// Add a file declaring a cross-import overlay.
   void addCrossImportOverlayFile(StringRef file);
 
+  /// Returns true if this module is known to not declare any cross-import
+  /// overlays. This doesn't actually load the files, so \c true may be
+  /// incorrect but \c false is never wrong.
```
Comment (brentdax) [inline include/swift/AST/Module.h]: Should be `\c false may be incorrect but \c true is never wrong`.

File: test/CrossImport/skip-submodules.swift [hunk 2]
```diff
@@ -4,6 +4,11 @@
 // RUN: %FileCheck %s < %t.txt
 // RUN: %FileCheck --check-prefix=NEGATIVE %s < %t.txt
 
+// FIXME: This test does not pass because another optimization is now skipping
+// this work entirely. It needs to be redesigned to use modules which have both
+// submodules *and* cross-import overlays.
+// XFAIL: *
```
Comment (brentdax) [inline test/CrossImport/skip-submodules.swift]: If someone does merge this, please file a radar on me and add the radar number here.

File: include/swift/AST/Module.h (hunk 1 above)
Comment (harlanhaskins) [reply include/swift/AST/Module.h]: Drive-by suggestion: would it help to make this a 2-state enum? NoCrossImportOverlays and MayHaveCrossImportOverlays?

File: lib/Sema/NameBinding.cpp [hunk 3]
```diff
@@ -902,55 +917,71 @@ void NameBinder::crossImport(ModuleDecl *M, UnboundImport &I) {
     // FIXME: Should we warn if M doesn't reexport underlyingModule?
     SF.addSeparatelyImportedOverlay(M, I.getUnderlyingModule().get());
 
-  // FIXME: Most of the comparisons we do here are probably unnecessary. We
-  // only need to findCrossImports() on pairs where at least one of the two
-  // modules declares cross-imports, and most modules don't. This is low-hanging
-  // performance fruit.
-
   auto newImports = crossImportableModules.getArrayRef()
                         .slice(nextModuleToCrossImport);
```
Comment (varungandhi-apple) [inline lib/Sema/NameBinding.cpp]: Not part of this PR but it would be better to use `drop_front` instead of `slice` as the latter's semantics are not clear from the name.

File: lib/Sema/NameBinding.cpp [hunk 4]
```diff
@@ -902,55 +917,71 @@ void NameBinder::crossImport(ModuleDecl *M, UnboundImport &I) {
...
+
+    // Fine, we need to do the slow-but-rare thing: check if this import
+    // declares a cross-import with any previous one.
     auto oldImports =
-        make_range(crossImportableModules.getArrayRef().data(), &newImport);
-    for (auto &oldImport : oldImports) {
-      if (!canCrossImport(oldImport))
-        continue;
+        llvm::makeArrayRef(crossImportableModules.getArrayRef().data(),
+                           &newImport);
```
Comment (varungandhi-apple) [inline lib/Sema/NameBinding.cpp]: We should probably annotate the arguments to indicate which overload is being picked. Replacing the `data()` call with `begin()` would help, as otherwise, one might think one is trying to pick the ptr+size overload.

File: lib/Sema/NameBinding.cpp [hunk 5]
```diff
@@ -978,6 +1017,11 @@ void NameBinder::findCrossImports(UnboundImport &I,
     unboundImports.emplace_back(declaringImport.module.second->getASTContext(),
                                 I, name, declaringImport, bystandingImport);
 
+    if (llvm::is_contained(oppositeNames, name))
```
Comment (varungandhi-apple) [inline lib/Sema/NameBinding.cpp]: I think we should default-initialize `oppositeNames` instead of leaving it uninitialized; otherwise, we might access uninitialized memory here?

File: include/swift/AST/Module.h (hunk 1 above)
Comment (brentdax) [reply include/swift/AST/Module.h]: I decided that's too heavyweight; instead, I'm flipping the sense of the return value and changing its name to `mightDeclareCrossImportOverlays()`, which accurately conveys the uncertainty.

File: lib/Sema/NameBinding.cpp (hunk 5 above)
Comment (brentdax) [reply lib/Sema/NameBinding.cpp]: A `SmallVector` is default-initialized without doing anything special—only primitive types (including pointers) are left uninitialized by default in C++.

Comment (varungandhi-apple) [reply lib/Sema/NameBinding.cpp]: Whaaaaat, that is so weird. I would've though leaving larger types uninitialized would be more useful than smaller types because there is more work that can be avoided. Sigh.

File: lib/Sema/NameBinding.cpp (hunk 4 above)
Comment (brentdax) [reply lib/Sema/NameBinding.cpp]: I didn't think the comments would really help here, but I added a comment explaining what we're creating.

**Comment by brentdax** [comment]
@swift-ci test

**Comment by brentdax** [comment]
@swift-ci test windows platform

**Comment by brentdax** [comment]
@swift-ci smoke test compiler performance

File: test/CrossImport/skip-submodules.swift (hunk 2 above)
Comment (brentdax) [reply test/CrossImport/skip-submodules.swift]: Deleted this test in favor of a more useful one.

**Comment by brentdax** [comment]
@swift-ci please smoke test linux platform
